src/plugins/fs.c

```diff
@@ -1486,39 +1571,142 @@ static gboolean xfs_resize_device (const gchar *device, guint64 new_size, const
-            return FALSE;
-        } else {
-            g_prefix_error (error, "Error when trying to detect filesystem on '%s': ", device);
-            return FALSE;
-        }
+/**
+ * bd_fs_repair:
+ * @device: the device the file system of which to repair
+ * @error: (out): place to store error (if any)
+ *
+ * Repair filesystem on @device. This calls other fs repair functions from this
+ * plugin based on detected filesystem (e.g. bd_fs_xfs_repair for XFS). This
+ * function will return an error for unknown/unsupported filesystems.
+ *
+ * Returns: whether the file system on @device was successfully repaired or not
+ */
+gboolean bd_fs_repair (const gchar *device, GError **error) {
+    return _do_operation (device, BD_FS_REPAIR, 0, error);
+}
+
+/**
+ * bd_fs_check:
+ * @device: the device the file system of which to check
+ * @error: (out): place to store error (if any)
+ *
+ * Check filesystem on @device. This calls other fs check functions from this
+ * plugin based on detected filesystem (e.g. bd_fs_xfs_check for XFS). This
+ * function will return an error for unknown/unsupported filesystems.
+ *
+ * Returns: whether the file system on @device passed the consistency check or not
+ */
+gboolean bd_fs_check (const gchar *device, GError **error) {
+    return _do_operation (device, BD_FS_CHECK, 0, error);
+}
+
+static gboolean _check_fs_operation (const gchar *fs_type, BDFsOpType op, gchar **required_utility, BDFsResizeFlags *mode, GError **error) {
+    gboolean ret;
+    const BDFSInfo *fsinfo = NULL;
+    const gchar* op_name = NULL;
+    guint64 exec_util_field = 0;
+    const gchar* exec_util = NULL;
+
+    switch (op) {
+        case BD_FS_RESIZE:
+            op_name = "Resizing";
+            exec_util_field = G_STRUCT_OFFSET (BDFSInfo, resize_util);
+            break;
+        case BD_FS_REPAIR:
+            op_name = "Repairing";
+            exec_util_field = G_STRUCT_OFFSET (BDFSInfo, repair_util);
+            break;
+        case BD_FS_CHECK:
+            op_name = "Checking";
+            exec_util_field = G_STRUCT_OFFSET (BDFSInfo, check_util);
+            break;
     }
 
-    if (g_strcmp0 (fstype, "ext2") == 0) {
-        return bd_fs_ext2_resize (device, new_size, NULL, error);
-    } else if (g_strcmp0 (fstype, "ext3") == 0) {
-        return bd_fs_ext3_resize (device, new_size, NULL, error);
-    } else if (g_strcmp0 (fstype, "ext4") == 0) {
-        return bd_fs_ext4_resize (device, new_size, NULL, error);
-    } else if (g_strcmp0 (fstype, "xfs") == 0) {
-        return xfs_resize_device (device, new_size, NULL, error);
-    } else if (g_strcmp0 (fstype, "vfat") == 0) {
-        return bd_fs_vfat_resize (device, new_size,  error);
-    } else {
+    if (required_utility != NULL)
+        *required_utility = NULL;
+
+    if (mode != NULL)
+        *mode = 0;
+
+    fsinfo = bd_fs_get_info (fs_type);
+    if (fsinfo != NULL)
+        exec_util = G_STRUCT_MEMBER (const gchar *, fsinfo, exec_util_field);
```

Comment:
I believe this function could be much less "magical" and avoid using the ``G_STRUCT`` macros if the steps were reordered -- i.e. get the ``fs_info`` first and then extract the utility from it in a ``switch``.